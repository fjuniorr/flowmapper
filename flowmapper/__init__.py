"""flowmapper."""

__all__ = (
    "__version__",
    # Add functions and variables you want exposed in `flowmapper.` namespace here
)

__version__ = "0.0.0-9008"
