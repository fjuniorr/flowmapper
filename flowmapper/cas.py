from functools import cached_property

class CAS():
    """
    Class for CAS Registry Numbers® (CAS RN®) that accepts padded or non-padded strings
    """
    def __init__(self, cas: str):
        if not isinstance(cas, str):
            raise TypeError(f'cas should be a str, not {type(cas).__name__}')
        else:
            self.cas = cas.strip().lstrip('0')
            self.digits = tuple(int(d) for d in self.cas[0:-1].replace('-', ''))
    def __repr__(self):
        return self.cas
    def __eq__(self, other):
        if not self.cas and not other.cas:
                result = False
        else:
            result = self.cas == other.cas
        return result
    @cached_property
    def check_digit_expected(self):
        """
        Expected digit acording to https://www.cas.org/support/documentation/chemical-substances/checkdig algorithm
        """
        result = sum([
            index * value 
            for index, value in enumerate(self.digits[::-1], start=1)
        ]) % 10
        return result
    @property
    def check_digit_actual(self):
        int(self.cas[-1])
    @property
    def valid(self):
        """
        True if check if CAS number is valid acording to https://www.cas.org/support/documentation/chemical-substances/checkdig algorithm
        """
        return self.check_digit_actual == self.check_digit_expected
