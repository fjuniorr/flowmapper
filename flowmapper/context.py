from dataclasses import asdict, dataclass

import flowmapper.jsonpath as jp


@dataclass
class Context:
    full: str = None
    primary: str = None
    secondary: str = None

    @classmethod
    def from_dict(cls, d, fields):
        fields = [fields] if isinstance(fields, str) else fields
        context_mapping_length = len(fields)
        primary_context = jp.extract(fields[0], d)
        if context_mapping_length == 1:
            result = Context(primary_context, primary_context)
        elif context_mapping_length == 2:
            secondary_context = jp.extract(fields[1], d)
            secondary_context = secondary_context if isinstance(secondary_context, str) else '/'.join(secondary_context)
            result = Context(
                full = f"{primary_context}/{secondary_context}",
                primary = primary_context,
                secondary = secondary_context
            )
        return result

    def to_dict(self):
        return asdict(self)
