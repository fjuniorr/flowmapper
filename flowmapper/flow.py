import flowmapper.jsonpath as jp

from .unit import Unit
from .cas import CAS
from .context import Context
from .utils import generate_flow_id, transform_flow
from .flowproperty import FlowProperty


class Flow:
    def __init__(
        self,
        original,
        fields=None,
        transformations=None,
    ):
        transformed = (
            transform_flow(original, transformations) if transformations else original
        )
        fields = fields if fields else {k:k for k in original}
        uuid_spec = fields.get("uuid")
        name_spec = fields.get("name")
        synonyms_spec = fields.get("synonyms")
        context_spec = fields.get("context")
        unit_spec = fields.get("unit")
        cas_spec = fields.get("cas")

        self.id = generate_flow_id(original)
        self.uuid = jp.extract(uuid_spec, transformed)
        self.uuid_raw_value = jp.extract(uuid_spec, original)
        self.uuid_raw_object = jp.extract(uuid_spec, original, "object")
        self.name = FlowProperty.from_dict(transformed, name_spec)
        self.name_raw_value = jp.extract(name_spec, original)
        self.name_raw_object = jp.extract(name_spec, original, "object")
        self.synonyms = FlowProperty.from_dict(transformed, synonyms_spec)
        self.synonyms_raw_value = jp.extract(synonyms_spec, original)
        self.synonyms_raw_object = jp.extract(synonyms_spec, original, "object")
        self.context = Context.from_dict(transformed, context_spec)
        self.context_raw_value = jp.extract(context_spec, original)
        self.context_raw_object = jp.extract(context_spec, original, "object")
        self.unit = Unit.from_dict(transformed, fields["unit"])
        self.unit_raw_value = jp.extract(unit_spec, original)
        self.unit_raw_object = jp.extract(unit_spec, original, "object")
        self.cas = CAS(jp.extract(cas_spec, transformed))
        self.cas_raw_value = jp.extract(cas_spec, original)
        self.cas_raw_object = jp.extract(cas_spec, original, "object")
        self.raw = original
        self.transformed = transformed

    def __repr__(self) -> str:
        return f"{self.name} (in {self.unit.value}) <{self.context.value}>"

    def __eq__(self, other):
        return self.id == other.id

    def __hash__(self):
        return hash(self.id)

    def __lt__(self, other):
        return self.name.raw_value < other.name.raw_value
