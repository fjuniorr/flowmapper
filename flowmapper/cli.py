import typer
from typing_extensions import Annotated
from pathlib import Path
import logging
from typing import Optional
from typing_extensions import Annotated
import importlib.metadata
import logging
import json
from pathlib import Path
from enum import Enum
from randonneur import migrate_datasets
from .flow import Flow
from .utils import read_field_mapping, read_flowlist, read_migration_files
from .flowmap import Flowmap
try:
    import tomllib
except ModuleNotFoundError:
    import tomli as tomllib

logger = logging.getLogger(__name__)

app = typer.Typer()

def version_callback(value: bool):
    if value:
        print(f"flowmapper, version {importlib.metadata.version('flowmapper')}")
        raise typer.Exit()

class OutputFormat(str, Enum):
    all = 'all'
    glad = 'glad'
    randonneur = 'randonneur'

@app.callback()
def main(version: Annotated[
        Optional[bool],
        typer.Option("--version", callback=version_callback, is_eager=True),
    ] = None,):
    """
    Generate mappings between elementary flows lists
    """

@app.command()
def map(
    source: Annotated[Path, typer.Argument(help='Path to source flowlist')],
    target: Annotated[Path, typer.Argument(help='Path to target flowlist')],
    fields: Annotated[Path, typer.Option(help='Relationship between fields in source and target flowlists')],
    output_dir: Annotated[Path, typer.Option(help='Directory to save mapping and diagnostics files')] = Path('.'),
    format: Annotated[OutputFormat, typer.Option(help='Mapping file output format', case_sensitive=False)] = 'all',
    transformations: Annotated[list[Path], typer.Option("--transformations", "-t", help='Randonneur data migration file with changes to be applied to source flows before matching')] = None,
    unmatched_source: Annotated[bool, typer.Option(help='Write original source unmatched flows into separate file?')] = True,
    unmatched_target: Annotated[bool, typer.Option(help='Write original target unmatched flows into separate file?')] = True,
    matched_source: Annotated[bool, typer.Option(help='Write original source matched flows into separate file?')] = False,
    matched_target: Annotated[bool, typer.Option(help='Write original target matched flows into separate file?')] = False,
):
    """
    Generate mappings between elementary flows lists
    """
    output_dir.mkdir(parents=True, exist_ok=True)
    field_mapping = read_field_mapping(fields)
    if transformations:
            transformations = read_migration_files(*transformations)
    source_flows = [Flow(flow, field_mapping['source'], transformations) for flow in read_flowlist(source)]
    target_flows = [Flow(flow, field_mapping['target']) for flow in read_flowlist(target)]

    flowmap = Flowmap(source_flows, target_flows)
    flowmap.statistics()

    stem = f'{source.stem}-{target.stem}'

    if matched_source:
        with open(output_dir / f'{stem}-matched-source.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.matched_source], fs, indent=True)

    if unmatched_source:
        with open(output_dir / f'{stem}-unmatched-source.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.unmatched_source], fs, indent=True)

    if matched_target:
        with open(output_dir / f'{stem}-matched-target.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.matched_target], fs, indent=True)

    if unmatched_target:
        with open(output_dir / f'{stem}-unmatched-target.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.unmatched_target], fs, indent=True)

    if format.value == 'randonneur':
        with open(output_dir / f'{stem}.json', 'w') as fs:
            json.dump(flowmap.to_randonneur(), fs, indent=2)
    elif format.value == 'glad':
        flowmap.to_glad().to_excel(output_dir / f'{stem}.xlsx', index = False)
    else:
        with open(output_dir / f'{stem}.json', 'w') as fs:
            json.dump(flowmap.to_randonneur(), fs, indent=2)
        flowmap.to_glad().to_excel(output_dir / f'{stem}.xlsx', index = False)
