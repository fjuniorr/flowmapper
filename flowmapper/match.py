import logging
from tqdm import tqdm

from .constants import (
    MINOR_LAND_NAME_DIFFERENCES_MAPPING,
    MISSING_FOSSIL_AND_BIOGENIC_CARBON_MAPPING,
    RANDOM_NAME_DIFFERENCES_MAPPING,
    NAME_DIFFERENCES_WITH_UNIT_CONVERSION_MAPPING,
)
from .flow import Flow
from .utils import rm_parentheses_roman_numerals, extract_country_code, rm_roman_numerals_ionic_state

logger = logging.getLogger(__name__)

def format_match_result(s: Flow, t: Flow, conversion_factor: float, match_info: dict):
    source_result = {
                     **s.name.raw_object, 
                     **s.context.raw_object,
                     **s.unit.raw_object
    }
    if s.uuid:
        source_result.update({s.fields['uuid']: s.uuid})
    
    target_result = {
                'uuid': t.uuid,
                'name': t.name.raw_value,
                'context': t.context.raw_value,
                'unit': t.unit.raw_value
            }
    if match_info.get('location'):
        target_result.update({'location': match_info['location']})

    result = {
            'source': source_result,
            'target': target_result,
            'conversion_factor': conversion_factor,
            'comment': match_info['comment']
        }
    return result

def match_identical_names_in_synonyms(s: Flow, t: Flow, comment: str = 'Identical synonyms'):
    is_match = True if t.synonyms and s.name.value in t.synonyms.value and s.context == t.context else False
    if is_match:
        return {'comment': comment}

def match_identical_cas_numbers(s: Flow, t: Flow, comment: str = 'Identical CAS numbers'):    
    is_match = s.cas == t.cas and s.context == t.context
    if is_match:
        return {'comment': comment}

def match_identical_names(s: Flow, t: Flow, comment = 'Identical names'):
    is_match = s.name == t.name and s.context == t.context
    
    if is_match:
        return {'comment': comment}

def match_identical_names_except_missing_suffix(s: Flow, t: Flow, suffix, comment = 'Identical names except missing suffix'):
    is_match = (
                (f"{s.name}, {suffix}" == t.name) or
                (f"{t.name}, {suffix}" == s.name) or
                (f"{s.name} {suffix}" == t.name) or
                (f"{t.name} {suffix}" == s.name)
    ) and s.context == t.context
    
    if is_match:
        return {'comment': comment}

def match_mapped_name_differences(s: Flow, t: Flow, mapping, comment = 'Mapped name differences'):    
    is_match = mapping.get(s.name) == t.name and s.context == t.context
    
    if is_match:
        return {'comment': comment}

def match_names_with_roman_numerals_in_parentheses(s: Flow, t: Flow, comment = 'With/without roman numerals in parentheses'):
    is_match = rm_parentheses_roman_numerals(s.name.value) == rm_parentheses_roman_numerals(t.name.value) and s.context == t.context
    
    if is_match:
        return {'comment': comment}

def match_names_with_country_codes(s: Flow, t: Flow, comment = 'Names with country code'):
    s_name, s_location = extract_country_code(s.name.value)
    is_match = s_location and s_name == t.name and s.context == t.context
    
    if is_match:
        return {'comment': comment, 'location': s_location.upper()}

def match_non_ionic_state(s: Flow, t: Flow, comment = 'Non-ionic state if no better match'):
    is_match = rm_roman_numerals_ionic_state(s.name.value) == t.name and s.context == t.context

    if is_match:
        return {'comment': comment}

def match_biogenic_to_non_fossil(s: Flow, t: Flow, comment = 'Biogenic to non-fossil if no better match'):
    is_match = s.name.value.removesuffix(', biogenic') == t.name.value.removesuffix(', non-fossil') and s.context == t.context

    if is_match:
        return {'comment': comment}

def match_resources_with_suffix_in_ground(s: Flow, t: Flow):
    return match_identical_names_except_missing_suffix(s, t, suffix = 'in ground', comment = 'Resources with suffix in ground')

def match_emissions_with_suffix_ion(s: Flow, t: Flow):
    return match_identical_names_except_missing_suffix(s, t, suffix = 'ion', comment = 'Match emissions with suffix ion')

def match_minor_random_name_differences(s: Flow, t: Flow):
    return match_mapped_name_differences(s, t, mapping = RANDOM_NAME_DIFFERENCES_MAPPING, comment = 'Minor random name differences')

def match_minor_land_name_differences(s: Flow, t: Flow):
    return match_mapped_name_differences(s, t, mapping = MINOR_LAND_NAME_DIFFERENCES_MAPPING, comment = 'Minor land name differences')

def match_missing_fossil_and_biogenic_carbon(s: Flow, t: Flow):
    return match_mapped_name_differences(s, t, mapping = MISSING_FOSSIL_AND_BIOGENIC_CARBON_MAPPING, comment = 'Missing biogenic and fossil carbon')

def match_mapped_name_differences_with_unit_conversion(s: Flow, t: Flow):
    return match_mapped_name_differences(s, t, mapping = NAME_DIFFERENCES_WITH_UNIT_CONVERSION_MAPPING, comment = 'Mapped name differences with unit conversions')

def match_rules(): 
    return [
            match_identical_names,
            match_identical_names_in_synonyms,
            match_resources_with_suffix_in_ground,
            match_minor_random_name_differences,
            match_emissions_with_suffix_ion,
            match_minor_land_name_differences,
            match_missing_fossil_and_biogenic_carbon,
            match_names_with_roman_numerals_in_parentheses,
            match_names_with_country_codes,
            match_mapped_name_differences_with_unit_conversion,
            match_identical_cas_numbers,
            match_non_ionic_state,
            match_biogenic_to_non_fossil,
    ]
