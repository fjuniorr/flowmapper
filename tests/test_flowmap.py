import pandas as pd
import json
from flowmapper import Flowmap
from flowmapper.match import (
    match_emissions_with_suffix_ion,
    match_identical_names,
)

def test_flowmap_remove_duplicates(source_flows, target_flows, snapshot):
    flowmap = Flowmap(source_flows, target_flows)
    actual = flowmap.source_flows
    assert actual == snapshot

def test_flowmap_mappings(source_flows, target_flows):
    flowmap = Flowmap(source_flows, target_flows)
    actual = flowmap.mappings[0]
    assert list(actual.keys()) == [
        "from",
        "to",
        "conversion_factor",
        "match_rule",
        "match_rule_priority",
        "info",
    ]
    assert actual["match_rule"] == "match_identical_names"


def test_flowmap_to_randonneur(source_flows, target_flows, snapshot):
    flowmap = Flowmap(source_flows, target_flows)
    actual = flowmap.to_randonneur()
    assert actual == snapshot

def test_flowmap_to_randonneur_export(source_flows, target_flows, snapshot, tmp_path):
    flowmap = Flowmap(source_flows, target_flows)
    flowmap.to_randonneur(tmp_path / "randonneur.json")
    with open(tmp_path / "randonneur.json", "r") as fs:
        actual = json.load(fs)
    assert actual == snapshot

def test_flowmap_with_custom_rules_no_match(source_flows, target_flows, snapshot):
    flowmap = Flowmap(
        source_flows,
        target_flows,
        rules=[match_emissions_with_suffix_ion],
    )
    actual = flowmap.mappings
    assert actual == snapshot


def test_flowmap_with_custom_rules_match(source_flows, target_flows, snapshot):
    flowmap = Flowmap(source_flows, target_flows, rules=[match_identical_names])
    actual = flowmap.to_randonneur()
    assert actual == snapshot


def test_flowmap_to_glad(source_flows, target_flows):
    flowmap = Flowmap(source_flows, target_flows)
    actual = flowmap.to_glad()
    expected = {
        "SourceFlowName": ["1,4-Butanediol", "Ammonia, FR"],
        "SourceFlowUUID": [None, None],
        "SourceFlowContext": ["Air/(unspecified)", "Emissions to air/low. pop."],
        "SourceUnit": ["kg", "kg"],
        "MatchCondition": ["", ""],
        "ConversionFactor": [1.0, 1.0],
        "TargetFlowName": ["1,4-Butanediol", "Ammonia"],
        "TargetFlowUUID": [
            "09db39be-d9a6-4fc3-8d25-1f80b23e9131",
            "0f440cc0-0f74-446d-99d6-8ff0e97a2444",
        ],
        "TargetFlowContext": [
            "air/unspecified",
            "air/non-urban air or from high stacks",
        ],
        "TargetUnit": ["kg", "kg"],
        "MemoMapper": ["Identical names", "Names with country code"],
    }
    assert actual.equals(pd.DataFrame(expected))

def test_flowmap_to_glad_export(source_flows, target_flows, tmp_path):
    flowmap = Flowmap(source_flows, target_flows)
    flowmap.to_glad(tmp_path / "glad.xlsx")
    actual = pd.read_excel(tmp_path / "glad.xlsx")
    expected = {
        "SourceFlowName": ["1,4-Butanediol", "Ammonia, FR"],
        "SourceFlowUUID": [float("nan"), float("nan")],
        "SourceFlowContext": ["Air/(unspecified)", "Emissions to air/low. pop."],
        "SourceUnit": ["kg", "kg"],
        "MatchCondition": [float("nan"), float("nan")],
        "ConversionFactor": [1, 1],
        "TargetFlowName": ["1,4-Butanediol", "Ammonia"],
        "TargetFlowUUID": [
            "09db39be-d9a6-4fc3-8d25-1f80b23e9131",
            "0f440cc0-0f74-446d-99d6-8ff0e97a2444",
        ],
        "TargetFlowContext": [
            "air/unspecified",
            "air/non-urban air or from high stacks",
        ],
        "TargetUnit": ["kg", "kg"],
        "MemoMapper": ["Identical names", "Names with country code"],
    }
    assert actual.equals(pd.DataFrame(expected))

def test_flowmap_export_matched(source_flows, target_flows, snapshot):
    flowmap = Flowmap(source_flows, target_flows)
    actual = [flow.raw for flow in flowmap.matched_source]
    assert actual == snapshot

def test_flowmap_export_unmatched(source_flows, target_flows, snapshot):
    flowmap = Flowmap(source_flows, target_flows)
    actual = [flow.raw for flow in flowmap.unmatched_source]
    assert actual == snapshot

def test_flowmap_nomatch_rule(source_flows, target_flows, snapshot):
    nomatch = lambda flow: flow.context.value == "air/urban air close to ground"
    flowmap = Flowmap(source_flows, target_flows, nomatch_rules=[nomatch])

    actual = [flow.raw for flow in flowmap.source_flows_nomatch]
    assert actual == snapshot

def test_flowmap_nomatch_rule_false(source_flows, target_flows, snapshot):
    nomatch = lambda flow: flow.context.value == "water"
    flowmap = Flowmap(source_flows, target_flows, nomatch_rules=[nomatch])

    actual = [flow.raw for flow in flowmap.source_flows]
    assert actual == snapshot

def test_flowmap_nomatch_multiple_rules(source_flows, target_flows, snapshot):
    nomatch1 = lambda flow: flow.context.value == "air/urban air close to ground"
    nomatch2 = lambda flow: flow.context.value == "air"
    flowmap = Flowmap(source_flows, target_flows, nomatch_rules=[nomatch1, nomatch2])

    actual = [flow.raw for flow in flowmap.source_flows_nomatch]
    assert actual == snapshot

def test_flowmap_mappings_ei_ei(target_flows, snapshot):
    flowmap = Flowmap(target_flows, target_flows)
    actual = flowmap.to_randonneur()
    assert actual == snapshot

def test_flowmap_mappings_ei39_ei310(flows_ei39, flows_ei310, snapshot):
    flowmap = Flowmap(flows_ei39, flows_ei310)
    actual = flowmap.to_randonneur()
    assert actual == snapshot
