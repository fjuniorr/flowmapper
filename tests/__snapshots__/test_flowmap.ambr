# serializer version: 1
# name: test_flowmap_export_matched
  list([
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        '(unspecified)',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Ammonia, FR',
      'unit': 'kg',
    }),
  ])
# ---
# name: test_flowmap_export_unmatched
  list([
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        'high. pop.',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Cesium-134',
      'unit': 'kBq',
    }),
    dict({
      'categories': list([
        'Emissions to soil',
        '',
      ]),
      'name': 'Cesium-134',
      'unit': 'kBq',
    }),
    dict({
      'categories': list([
        'Resources',
        'in ground',
      ]),
      'name': 'Zinc, Zn 0.63%, Au 9.7E-4%, Ag 9.7E-4%, Cu 0.38%, Pb 0.014%, in ore',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Ammonia, as N',
      'unit': 'kg',
    }),
  ])
# ---
# name: test_flowmap_mappings_ei39_ei310
  list([
    dict({
      'comment': 'Identical uuid',
      'conversion_factor': 1.0,
      'source': dict({
        '@id': '4f777e05-70f9-4a18-a406-d8232325073f',
        'compartment': dict({
          '@subcompartmentId': 'be7e06e9-0bf5-462e-99dc-fe4aee383c48',
          'compartment': dict({
            '#text': 'air',
            '@xml:lang': 'en',
          }),
          'subcompartment': dict({
            '#text': 'non-urban air or from high stacks',
            '@xml:lang': 'en',
          }),
        }),
        'name': dict({
          '#text': '2,4-D amines',
          '@xml:lang': 'en',
        }),
        'unitName': dict({
          '#text': 'kg',
          '@xml:lang': 'en',
        }),
      }),
      'target': dict({
        'context': 'air/non-urban air or from high stacks',
        'name': '2,4-D dimethylamine salt',
        'unit': 'kg',
        'uuid': 'b6b4201e-0561-5992-912f-e729fbf04e41',
      }),
    }),
  ])
# ---
# name: test_flowmap_mappings_ei_ei
  list([
    dict({
      'comment': 'Identical uuid',
      'conversion_factor': 1.0,
      'source': dict({
        '@id': '09db39be-d9a6-4fc3-8d25-1f80b23e9131',
        'compartment': dict({
          '@subcompartmentId': '7011f0aa-f5f9-4901-8c10-884ad8296812',
          'compartment': dict({
            '#text': 'air',
            '@xml:lang': 'en',
          }),
          'subcompartment': dict({
            '#text': 'unspecified',
            '@xml:lang': 'en',
          }),
        }),
        'name': dict({
          '#text': '1,4-Butanediol',
          '@xml:lang': 'en',
        }),
        'unitName': dict({
          '#text': 'kg',
          '@xml:lang': 'en',
        }),
      }),
      'target': dict({
        'context': 'air/unspecified',
        'name': '1,4-Butanediol',
        'unit': 'kg',
        'uuid': '09db39be-d9a6-4fc3-8d25-1f80b23e9131',
      }),
    }),
    dict({
      'comment': 'Identical uuid',
      'conversion_factor': 1.0,
      'source': dict({
        '@id': '0f440cc0-0f74-446d-99d6-8ff0e97a2444',
        'compartment': dict({
          '@subcompartmentId': 'be7e06e9-0bf5-462e-99dc-fe4aee383c48',
          'compartment': dict({
            '#text': 'air',
            '@xml:lang': 'en',
          }),
          'subcompartment': dict({
            '#text': 'non-urban air or from high stacks',
            '@xml:lang': 'en',
          }),
        }),
        'name': dict({
          '#text': 'Ammonia',
          '@xml:lang': 'en',
        }),
        'unitName': dict({
          '#text': 'kg',
          '@xml:lang': 'en',
        }),
      }),
      'target': dict({
        'context': 'air/non-urban air or from high stacks',
        'name': 'Ammonia',
        'unit': 'kg',
        'uuid': '0f440cc0-0f74-446d-99d6-8ff0e97a2444',
      }),
    }),
  ])
# ---
# name: test_flowmap_nomatch_multiple_rules
  list([
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        '(unspecified)',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        'high. pop.',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
  ])
# ---
# name: test_flowmap_nomatch_rule
  list([
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        'high. pop.',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
  ])
# ---
# name: test_flowmap_nomatch_rule_false
  list([
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        '(unspecified)',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        'high. pop.',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Cesium-134',
      'unit': 'kBq',
    }),
    dict({
      'categories': list([
        'Emissions to soil',
        '',
      ]),
      'name': 'Cesium-134',
      'unit': 'kBq',
    }),
    dict({
      'categories': list([
        'Resources',
        'in ground',
      ]),
      'name': 'Zinc, Zn 0.63%, Au 9.7E-4%, Ag 9.7E-4%, Cu 0.38%, Pb 0.014%, in ore',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Ammonia, FR',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Ammonia, as N',
      'unit': 'kg',
    }),
  ])
# ---
# name: test_flowmap_remove_duplicates
  list([
    1,4-Butanediol (in kg) <Air/(unspecified)> [uuid:None],
    1,4-Butanediol (in kg) <Air/high. pop.> [uuid:None],
    Cesium-134 (in kBq) <Emissions to air/low. pop.> [uuid:None],
    Cesium-134 (in kBq) <Emissions to soil/> [uuid:None],
    Zinc, Zn 0.63%, Au 9.7E-4%, Ag 9.7E-4%, Cu 0.38%, Pb 0.014%, in ore (in kg) <Resources/in ground> [uuid:None],
    Ammonia, FR (in kg) <Emissions to air/low. pop.> [uuid:None],
    Ammonia, as N (in kg) <Emissions to air/low. pop.> [uuid:None],
  ])
# ---
# name: test_flowmap_to_randonneur
  list([
    dict({
      'comment': 'Identical names',
      'conversion_factor': 1.0,
      'source': dict({
        'categories': list([
          'Air',
          '(unspecified)',
        ]),
        'name': '1,4-Butanediol',
        'unit': 'kg',
      }),
      'target': dict({
        'context': 'air/unspecified',
        'name': '1,4-Butanediol',
        'unit': 'kg',
        'uuid': '09db39be-d9a6-4fc3-8d25-1f80b23e9131',
      }),
    }),
    dict({
      'comment': 'Names with country code',
      'conversion_factor': 1.0,
      'source': dict({
        'categories': list([
          'Emissions to air',
          'low. pop.',
        ]),
        'name': 'Ammonia, FR',
        'unit': 'kg',
      }),
      'target': dict({
        'context': 'air/non-urban air or from high stacks',
        'location': 'FR',
        'name': 'Ammonia',
        'unit': 'kg',
        'uuid': '0f440cc0-0f74-446d-99d6-8ff0e97a2444',
      }),
    }),
  ])
# ---
# name: test_flowmap_to_randonneur_export
  list([
    dict({
      'comment': 'Identical names',
      'conversion_factor': 1.0,
      'source': dict({
        'categories': list([
          'Air',
          '(unspecified)',
        ]),
        'name': '1,4-Butanediol',
        'unit': 'kg',
      }),
      'target': dict({
        'context': 'air/unspecified',
        'name': '1,4-Butanediol',
        'unit': 'kg',
        'uuid': '09db39be-d9a6-4fc3-8d25-1f80b23e9131',
      }),
    }),
    dict({
      'comment': 'Names with country code',
      'conversion_factor': 1.0,
      'source': dict({
        'categories': list([
          'Emissions to air',
          'low. pop.',
        ]),
        'name': 'Ammonia, FR',
        'unit': 'kg',
      }),
      'target': dict({
        'context': 'air/non-urban air or from high stacks',
        'location': 'FR',
        'name': 'Ammonia',
        'unit': 'kg',
        'uuid': '0f440cc0-0f74-446d-99d6-8ff0e97a2444',
      }),
    }),
  ])
# ---
# name: test_flowmap_with_custom_rules_match
  list([
    dict({
      'comment': 'Identical names',
      'conversion_factor': 1.0,
      'source': dict({
        'categories': list([
          'Air',
          '(unspecified)',
        ]),
        'name': '1,4-Butanediol',
        'unit': 'kg',
      }),
      'target': dict({
        'context': 'air/unspecified',
        'name': '1,4-Butanediol',
        'unit': 'kg',
        'uuid': '09db39be-d9a6-4fc3-8d25-1f80b23e9131',
      }),
    }),
  ])
# ---
# name: test_flowmap_with_custom_rules_no_match
  list([
  ])
# ---
